File: HLPProject/HLPProject/State.fs
// Module that defines the instruction types and state

namespace Common
module State =
    
    (*** STATE ***)

    /// Represents the Machine State.
    type StateHandle = S of Regs:int [] * N:bool * Z:bool * C:bool * S:bool * Mem:Map<int,int>

    // Functions to access and modify state.

    /// Creates default state.
    let initState =
        let regs = Array.create 16 0
        S(regs, false, false, false, false, Map.empty)

    /// Read a register in the state.
    let readReg r (S(reg,_,_,_,_,_): StateHandle) =
        reg.[r]

    /// Write a value to a register in the state.
    let writeReg r v (S(reg,n,z,c,s,mem): StateHandle) =
        let newRegs = Array.mapi (fun i x -> if r = i then v else x) reg
        S(newRegs,n,z,c,s,mem)

    /// Write a value to a register in the state and update status flags.
    (*let writeRegFlags r v r1 r2 (S(reg,n,z,c,s,mem): StateHandle) =
        let newRegs = Array.mapi (fun i x -> if r = i then v else x) reg
        match v with
        | x when (x < r1) && (x < r2) ->  *)

    /// Read the value in the Program Counter.
    let readPC (S(reg,_,_,_,_,_): StateHandle) =
        reg.[15]

    /// Write a value to the Program Counter.
    let writePC v (S(reg,n,z,c,s,mem): StateHandle) =
        let newRegs = Array.mapi (fun i x -> if i = 15 then v else x) reg
        S(newRegs,n,z,c,s,mem)

    /// Increment the Program Counter by 4.
    let incPC (S(reg,n,z,c,s,mem): StateHandle) =
        let newRegs = Array.mapi (fun i x -> if i = 15 then x+4 else x) reg
        S(newRegs,n,z,c,s,mem)

    /// Retrieve negative flag.
    let readNFlag (S(_,n,_,_,_,_): StateHandle) = n

    /// Retrieve zero flag.
    let readZFlag (S(_,_,z,_,_,_): StateHandle) = z

    /// Retrieve carry flag.
    let readCFlag (S(_,_,_,c,_,_): StateHandle) = c
    
    /// Retrieve overflow flag.
    let readVFlag (S(_,_,_,_,v,_): StateHandle) = v
    
    /// Write negative flag.
    let writeNFlag n (S(reg,_,z,c,v,mem): StateHandle) =
        S(reg,n,z,c,v,mem)

    /// Write zero flag.
    let writeZFlag z (S(reg,n,_,c,v,mem): StateHandle) =
        S(reg,n,z,c,v,mem)

    /// Write carry flag.
    let writeCFlag c (S(reg,n,z,_,v,mem): StateHandle) =
        S(reg,n,z,c,v,mem)
    
    /// Write overflow flag.
    let writeVFlag v (S(reg,n,z,c,_,mem): StateHandle) =
        S(reg,n,z,c,v,mem)

    /// Read from a memory address.
    let readMem addr (S(_,_,_,_,_,mem): StateHandle) =
        match Map.tryFind addr mem with
        | Some(v) -> v
        | None -> 0

    /// Write to a memory address.
    let writeMem addr v (S(reg,n,z,c,s,mem): StateHandle) =
        let newMem = Map.add addr v mem
        S(reg,n,z,c,s,newMem)

File: HLPProject/HLPProject/hlp/State.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

import { setType } from "fable-core/Symbol";
import _Symbol from "fable-core/Symbol";
import { compare, compareUnions, equalsUnions, makeGeneric, Array as _Array } from "fable-core/Util";
import { add, tryFind, create } from "fable-core/Map";
import _Map from "fable-core/Map";
import { mapIndexed, replicate } from "fable-core/Seq";
import GenericComparer from "fable-core/GenericComparer";
export var StateHandle = function () {
  function StateHandle(caseName, fields) {
    _classCallCheck(this, StateHandle);

    this.Case = caseName;
    this.Fields = fields;
  }

  _createClass(StateHandle, [{
    key: _Symbol.reflection,
    value: function () {
      return {
        type: "Common.State.StateHandle",
        interfaces: ["FSharpUnion", "System.IEquatable", "System.IComparable"],
        cases: {
          S: [_Array(Int32Array, true), "boolean", "boolean", "boolean", "boolean", makeGeneric(_Map, {
            Key: "number",
            Value: "number"
          })]
        }
      };
    }
  }, {
    key: "Equals",
    value: function (other) {
      return equalsUnions(this, other);
    }
  }, {
    key: "CompareTo",
    value: function (other) {
      return compareUnions(this, other);
    }
  }]);

  return StateHandle;
}();
setType("Common.State.StateHandle", StateHandle);
export var initState = function () {
  var regs = Int32Array.from(replicate(16, 0));
  return new StateHandle("S", [regs, false, false, false, false, create(null, new GenericComparer(compare))]);
}();
export function readReg(r, _arg1) {
  return _arg1.Fields[0][r];
}
export function writeReg(r, v, _arg1) {
  var newRegs = Int32Array.from(mapIndexed(function (i, x) {
    return r === i ? v : x;
  }, _arg1.Fields[0]));
  return new StateHandle("S", [newRegs, _arg1.Fields[1], _arg1.Fields[2], _arg1.Fields[3], _arg1.Fields[4], _arg1.Fields[5]]);
}
export function readPC(_arg1) {
  return _arg1.Fields[0][15];
}
export function writePC(v, _arg1) {
  var newRegs = Int32Array.from(mapIndexed(function (i, x) {
    return i === 15 ? v : x;
  }, _arg1.Fields[0]));
  return new StateHandle("S", [newRegs, _arg1.Fields[1], _arg1.Fields[2], _arg1.Fields[3], _arg1.Fields[4], _arg1.Fields[5]]);
}
export function incPC(_arg1) {
  var newRegs = Int32Array.from(mapIndexed(function (i, x) {
    return i === 15 ? 4 : x;
  }, _arg1.Fields[0]));
  return new StateHandle("S", [newRegs, _arg1.Fields[1], _arg1.Fields[2], _arg1.Fields[3], _arg1.Fields[4], _arg1.Fields[5]]);
}
export function readNFlag(_arg1) {
  return _arg1.Fields[1];
}
export function readZFlag(_arg1) {
  return _arg1.Fields[2];
}
export function readCFlag(_arg1) {
  return _arg1.Fields[3];
}
export function readVFlag(_arg1) {
  return _arg1.Fields[4];
}
export function writeNFlag(n, _arg1) {
  return new StateHandle("S", [_arg1.Fields[0], n, _arg1.Fields[2], _arg1.Fields[3], _arg1.Fields[4], _arg1.Fields[5]]);
}
export function writeZFlag(z, _arg1) {
  return new StateHandle("S", [_arg1.Fields[0], _arg1.Fields[1], z, _arg1.Fields[3], _arg1.Fields[4], _arg1.Fields[5]]);
}
export function writeCFlag(c, _arg1) {
  return new StateHandle("S", [_arg1.Fields[0], _arg1.Fields[1], _arg1.Fields[2], c, _arg1.Fields[4], _arg1.Fields[5]]);
}
export function writeVFlag(v, _arg1) {
  return new StateHandle("S", [_arg1.Fields[0], _arg1.Fields[1], _arg1.Fields[2], _arg1.Fields[3], v, _arg1.Fields[5]]);
}
export function readMem(addr, _arg1) {
  var matchValue = tryFind(addr, _arg1.Fields[5]);

  if (matchValue == null) {
    return 0;
  } else {
    return matchValue;
  }
}
export function writeMem(addr, v, _arg1) {
  var newMem = add(addr, v, _arg1.Fields[5]);
  return new StateHandle("S", [_arg1.Fields[0], _arg1.Fields[1], _arg1.Fields[2], _arg1.Fields[3], _arg1.Fields[4], newMem]);
}